lib/attributor/hash_dsl_compiler.rb

```diff
@@ -0,0 +1,141 @@
+require_relative 'dsl_compiler'
+
+
+module Attributor
+
+  class HashDSLCompiler < DSLCompiler
+
+    # A class that encapsulates the definition of a requirement for Hash attributes
+    # It implements the validation against incoming values and it describes its format for documentation purposes
+    class Requirement
```

Comment:
It seems like this could be broken down into an abstract base Requirement, with subclasses for the specific types. 

Might we want to add, or support adding, extra types of requirements? Could this replace the fuzzy existing `requires_if` option, along with AttributeResolver and all its baggage?